File: Programmers/Algorithm_Kit/03_Heap/002.cpp
#include <string>
#include <vector>
#include <iostream>
#include <queue>
#include <algorithm> //sort
using namespace std;

class cmp {
public:
    bool operator() (pair<int, int> a, pair<int, int> b) {
        if (a.second == b.second) {
            return a.first > b.first;
        }

        return a.second > b.second; //오름차순 = pq greater = 최소 힙
    }
};

priority_queue<pair<int, int>, vector<pair<int, int>>, cmp> pq;

int solution(vector<vector<int>> jobs) {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    int answer = 0;

    sort(jobs.begin(), jobs.end());

    int time = 0;
    int idx = 0;

    while (idx < jobs.size() || !pq.empty()) {
        while (idx < jobs.size() && jobs[idx][0] <= time) {
            pq.push({ jobs[idx][0], jobs[idx][1] });
            idx++;
        }

        if (!pq.empty()) {
            pair<int, int> cur = pq.top(); pq.pop();

            time += cur.second;
            answer += time - cur.first;
        }
        else {
            time = jobs[idx][0];  //다음 작업의 시작 시간으로 점프
        }
    }

    return answer / jobs.size();
}

File: programmers/algorithmkit/03_heap/002.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        int answer = 0;

        Arrays.sort(jobs, (j1, j2) -> Integer.compare(j1[0], j2[0]));

        int t = 0;
        int idx = 0;

        PriorityQueue<Pair> pq = new PriorityQueue<>();

        while (idx < jobs.length || !pq.isEmpty()) {
            while (idx < jobs.length && jobs[idx][0] <= t) {
                pq.offer(new Pair(jobs[idx][0], jobs[idx][1]));
                idx++;
            }

            if (!pq.isEmpty()) {
                Pair cur = pq.poll();

                t += cur.time;
                answer += t - cur.start;
            } else {
                t = jobs[idx][0];
            }
        }

        return answer / jobs.length;
    }

    class Pair implements Comparable<Pair> {
        int start;
        int time;

        public Pair(int start, int time) {
            this.start = start;
            this.time = time;
        }

        @Override
        public int compareTo(Pair other) {
            if (this.time == other.time) {
                return Integer.compare(this.start, other.start);
            }

            return Integer.compare(this.time, other.time);
        }
    }
}
